File: app/src/main/java/com/dialogapp/dialog/ui/common/BaseListViewModel.java
package com.dialogapp.dialog.ui.common;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import com.dialogapp.dialog.model.Item;
import com.dialogapp.dialog.util.Resource;

import java.util.List;

/**
 * Base ViewModel class for list views
 */

public class BaseListViewModel extends ViewModel {
    protected final MutableLiveData<Boolean> refresh = new MutableLiveData<>();

    protected LiveData<Resource<List<Item>>> posts;

    public void refresh() {
        refresh.setValue(true);
    }

    public LiveData<Resource<List<Item>>> getPosts() {
        return posts;
    }
}

File: app/src/main/java/com/dialogapp/dialog/ui/common/BaseListViewModel.kt
package com.dialogapp.dialog.ui.common

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import com.dialogapp.dialog.model.Post
import com.dialogapp.dialog.repository.PostsRepository
import com.dialogapp.dialog.util.AbsentLiveData
import com.dialogapp.dialog.vo.CONVERSATION
import com.dialogapp.dialog.vo.DISCOVER
import com.dialogapp.dialog.vo.FAVORITES
import com.dialogapp.dialog.vo.Listing
import com.dialogapp.dialog.vo.Resource
import javax.inject.Inject

class BaseListViewModel @Inject constructor(private val postsRepository: PostsRepository)
    : ViewModel() {

    private val _endpoint = MutableLiveData<EndpointArgs>()
    val endpoint: LiveData<EndpointArgs>
        get() = _endpoint
    val endpointResult: LiveData<Resource<Listing<Post>>> = Transformations
            .switchMap(_endpoint) { endpointArgs ->
                when {
                    endpointArgs.endpoint == DISCOVER ->
                        postsRepository.loadDiscover(endpointArgs.extraArg)
                    endpointArgs.endpoint == FAVORITES ->
                        AbsentLiveData.create()
                    endpointArgs.endpoint == CONVERSATION ->
                        postsRepository.loadConversation(endpointArgs.extraArg ?: "")
                    else -> postsRepository.loadPostsByUsername(endpointArgs.endpoint)
                }
            }

    fun showEndpoint(endpointArgs: EndpointArgs) {
        if (_endpoint.value != endpointArgs) {
            _endpoint.value = endpointArgs
        }
    }

    fun refresh() {
        _endpoint.value?.let {
            _endpoint.value = it
        }
    }
}

data class EndpointArgs(
        val endpoint: String,
        val extraArg: String? = null
)